### ISSUE TITLE
[codeql-go]: Mass Add Web Framework Models With the Help of Code Generation

### ISSUE THREAD

**Issue by gagliardetto** [github/securitylab#335]
## Query

- https://github.com/github/codeql-go/pull/438
- https://github.com/github/codeql-go/pull/538
- https://github.com/github/codeql-go/pull/539

## CVE ID(s)

No CVEs, yet.

## Report

One of the main research areas of CodeQL queries are **web** technologies.

More often than not, those queries target scenarios where **user-provided data** ends up in **unwanted places**, opening the doors to vulnerabilities such as XSS, SQLi, SSRF, RCE, etc.

One of the obstacles to **finding all variants of those vulnerabilities** and **eradicating them forever** is that there are dozens of popular web frameworks for any given language ([and Go has a few, too](https://github.com/search?q=language%3AGo+web+framework)).

Wouldn't it be awesome if we could **model them all**? How many new **TPs** would that unlock? How many **CVEs** would that bring?

The *bad news* is that to model a web framework and write tests can take *days*.

The *good news* is that when there are **a few common patterns** that a group of libraries follow, that's where **automation** can shine the most (such is the case for web frameworks).

And following that logic, I rewrote my previous library (see https://github.com/github/securitylab/issues/187) and created [codemill](https://github.com/gagliardetto/codemill).

[Codemill](https://github.com/gagliardetto/codemill) is a tool with a graphical UI that can be **extended** to help with the creation of **many model kinds**; right now it covers these:

- **TaintTracking** - DONE
- **UntrustedFlowSource** - DONE
- **HTTP::HeaderWrite** - (Needs feedback on the model)
- **HTTP::Redirect** - (Needs feedback on the model)
- **HTTP::ResponseBody** - (Needs feedback on the model)

The above group was specifically written thinking about web frameworks; other kinds (such as `SystemCommandExecution`, `Logger` etc.) could be done next.

Here you can see it in action for an `UntrustedFlowSource` model:

![codemill-gin-untrustedflowsource](https://user-images.githubusercontent.com/15271561/113894302-8aa51b00-97d0-11eb-9645-c2a4f1d26958.gif)
 
codemill will generate CodeQL and Go test code.

With the semi-automatic approach of codemill, a typical web framework can be modeled in :clock11: __**less than an hour**__ :clock12: .

Here's a list of Go web frameworks, sorted by star count: [web frameworks by star count](https://github.com/gagliardetto/codemill/blob/main/lists/go/web-frameworks/by-stars.txt)

And here's a list of currently supported frameworks in CodeQL: https://codeql.github.com/docs/codeql-overview/supported-languages-and-frameworks/#go-built-in-support

- [ ] Are you planning to discuss this vulnerability submission publicly? (Blog Post, social networks, etc). *We would love to have you spread the word about the good work you are doing*

## Result(s)

*Provide at least one useful result found by your query, on some revision of a real project.*

**TODO**

**Comment by kevinbackhouse** [comment]
@gagliardetto: Just to confirm, this is still a work-in-progress, right? We will review it when it is ready.

**Comment by gagliardetto** [comment]
Yep, still in progress. Improving the generated models before going all in
on the frameworks.

**Comment by ghsecuritylab** [comment]
Your submission is now in status **SecLab review**.

_For information, the evaluation workflow is the following:
SecLab review > FP Check > CodeQL review > SecLab finalize > Pay > Closed_

**Comment by antonio-morales** [comment]
Hi @gagliardetto 

any update on this?

**Comment by gagliardetto** [comment]
Hi @antonio-morales

I'd like to know more about how to proceed with this specific bounty
request, given the recent changes to the policy. @xcorail

**Comment by xcorail** [comment]
👋🏾 @gagliardetto 

Re: changes to the policy
- we were already asking for CVE in the issue template, but it was not clear in the rules and not enforced
- this submission is prior to the clarification so we'll not retroactively enforce the CVE requirement, but we'll still enforce the "true positive" requirement

Re: specific bounty
- This is indeed a very specific submission, and we will review it on an exception basis, as the common rules do not really cover this case :-)

**Comment by yo-h** [comment]
@gagliardetto FYI we plan to switch Go models to use the approach for Java in [ExternalFlow.qll](https://github.com/github/codeql/blob/main/java/ql/src/semmle/code/java/dataflow/ExternalFlow.qll) and convert existing models in the next couple of months or so. You may wish to pause this submission until that is completed. Let us know if you have any questions.

**Comment by yo-h** [comment]
cc @smowton @owen-mc @sauyon

**Comment by smowton** [comment]
Worth noting that that only relates to taint-propagating function models and straightforward sources and sinks; I imagine for the foreseeable future things like noting which functions constitute HTTP header setters and what response body they correspond to will stay specified in QL.

**Comment by JarLob** [comment]
As I understand one last PR is left. @gagliardetto any progress on it?

**Comment by gagliardetto** [comment]
thanks for the ping, @JarLob 

I am no longer working on this, and don't plan on resuming.

**Comment by smowton** [comment]
Did you abandon it because of problems our end, or something else?

**Comment by gagliardetto** [comment]
You guys were amazing! I'm very grateful for all your reviews and lessons :hugs:

It's been a great pleasure to work with you all :vulcan_salute: 

The reason for abandoning this specific project was two mistakes on my part: I've underestimated the time it would take to complete it (my February 2021 estimate was **very off** :rofl: ), and I've overestimated the impact it would have and the interest it would spark.

Over the past few months my time and priorities have shifted towards other projects that don't involve go+codeql, and so I am reassessing my commitments and making the necessary cuts.

Codeql is awesome, and I have another project in the works that's intimately entangled with it, but not for Go.

Thanks y'all

Cheers!
